### PR TITLE
[DO NOT MERGE] Used tf.keras.backend.rnn to implement keras.backend.rnn

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
### Summary

This PR is another way of doing #11502
There is currently no global variable used in `tf.keras.backend.rnn` so it should be in theory a safe thing to do. There are some gotchas though.

In `keras.backend.rnn`, when executing the function, the attribute `_uses_learning_phase` of keras tensors was used a lot to determine what the attribute `_uses_learning_phase` of the last output should be. 

This attribute is not public and it is not used anywhere in the implementation of `tf.keras.backend.rnn`. So this might be our biggest source of issues.

Some investigation:
* When copying the keras codebase in tensorflow, this attribute was used in a few places but not in `K.rnn`.
* The usage of the attribute `_uses_learning_phase` was introduced in #7943 when adding support for rnn cells.

### Related Issues

#11502
#12170
#11720

### PR Overview

- [ ] This PR requires new unit tests [y/n] (make sure tests are included)
- [ ] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y/n] (I hope)
- [ ] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

**Comment by gabrieldemarmiesse** [comment]
@andhus Your input would be welcome on this if you have the time to spare.

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -3037,203 +3037,14 @@ def rnn(step_function, inputs, initial_states,
...
-        last_time = final_outputs[0]
-        output_ta = final_outputs[1]
-        outputs = output_ta.stack()
-        last_output = output_ta.read(last_time - 1)
-
-    axes = [1, 0] + list(range(2, len(outputs.shape)))
-    outputs = tf.transpose(outputs, axes)
-    last_output._uses_learning_phase = uses_learning_phase[0]
-    return last_output, outputs, new_states
+    return tf.keras.backend.rnn(step_function,
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Prefer using explicit keyword arguments, for clarity

**Comment by fchollet** [comment]
> _uses_learning_phase

The reasonable thing to do is to scrub this attribute from the codebase and unconditionally pass the learning phase value 1 to the learning phase placeholder in training mode. In inference mode we don't actually need to pass a value (at least in TF) because we can rely on the default value of the placeholder (0).

This is easy in TF (thanks to placeholder_with_default). Seems doable to support the same in other backends as well (the default injection can be done in `K.function`).

Overall `_uses_learning_phase` added a lot of brittle complexity and we are better off without it.

**Comment by gabrieldemarmiesse** [comment]
> Overall _uses_learning_phase added a lot of brittle complexity and we are better off without it.

I'm not really familiar with this attribute and what it is used for. So i'm not sure I'll be able to do the modifications you would like me to do.

I'll close this PR for now. I might take a stab at it again in the future but it's not certain.
